File: desktop/src/SmartSocket.App/Transport/BleTransport.cs
using System.Text;
using SmartSocket.Core;
using Windows.Devices.Bluetooth;
using Windows.Devices.Bluetooth.GenericAttributeProfile;
using Windows.Devices.Enumeration;
using Windows.Security.Cryptography;
using Windows.Storage.Streams;

namespace SmartSocket.App.Transport;

/// <summary>
/// The link to a Bluetooth Low Energy module.
///
/// WHY THIS EXISTS. "Bluetooth" is two incompatible radios. Classic (BR/EDR)
/// has SPP, which Windows exposes as a COM port, and that is what the HC-05 was
/// assumed to be. BLE has no SPP and no COM port at all - every vendor bolted a
/// serial channel onto GATT with private UUIDs. A BLE module can therefore be
/// paired, working, and completely invisible to a port list, which is precisely
/// what this project ran into: Windows reported a "Bluetooth LE Generic
/// Attribute Service" while every COM port on the machine belonged to something
/// else.
///
/// Which characteristics carry the bytes is worked out by
/// <see cref="BleSerialProfiles"/>, in the testable core, so this class only has
/// to deal with the radio.
/// </summary>
public sealed class BleTransport : ISocketTransport
{
    /// <summary>
    /// Long enough for a module to finish GATT discovery and answer, short
    /// enough that a wrong device is reported rather than hung on.
    /// </summary>
    private static readonly TimeSpan HandshakeTimeout = TimeSpan.FromSeconds(10);

    /// <summary>
    /// How many times to re-enumerate before believing a device really has no
    /// serial service. Three covers the settling window without making a
    /// genuinely wrong device take a minute to reject.
    /// </summary>
    private const int DiscoveryAttempts = 3;

    private BluetoothLEDevice? _device;
    private GattCharacteristic? _notify;
    private GattCharacteristic? _write;

    /// <summary>
    /// Held open deliberately. Without MaintainConnection, Windows drops a BLE
    /// link the moment it thinks nothing needs it - which is why the socket
    /// would connect, work, and then be gone a minute later for no visible
    /// reason.
    /// </summary>
    private GattSession? _session;

    /// <summary>
    /// THESE MUST BE KEPT ALIVE. GattDeviceService is IDisposable, and when the
    /// last reference to one is collected Windows tears down the GATT session
    /// underneath it - taking the characteristics and the connection with it.
    /// Holding only the characteristics, as this class first did, means the link
    /// dies at whatever arbitrary moment the garbage collector next runs.
    /// </summary>
    private readonly List<GattDeviceService> _services = [];

    private readonly StringBuilder _line = new();
    private readonly object _gate = new();

    private volatile bool _sawBytes;
    private volatile bool _sawStatus;
    private volatile bool _closing;

    public event Action<LinkState>? LinkStateChanged;
    public event Action<SocketStatus>? StatusChanged;

    /// <summary>Which profile matched, for the error messages.</summary>
    public string? ProfileName { get; private set; }

    /// <summary>
    /// Every BLE device Windows has paired. Deliberately paired-only: scanning
    /// would need a live radio sweep and would list every fitness tracker in
    /// the building, and the socket has to be paired to be useful anyway.
    /// </summary>
    public static async Task<IReadOnlyList<SocketDevice>> PairedDevicesAsync()
    {
        try
        {
            var selector = BluetoothLEDevice.GetDeviceSelectorFromPairingState(true);
            var found = await DeviceInformation.FindAllAsync(selector);

            return found
                .Select(d => new SocketDevice(
                    string.IsNullOrWhiteSpace(d.Name) ? "Unnamed BLE device" : d.Name,
                    d.Id,
                    LinkKind.Ble))
                .ToList();
        }
        catch (Exception)
        {
            // No radio, or the adapter is off. An empty list is the honest
            // answer and the caller shows only the COM ports.
            return [];
        }
    }

    public async Task<bool> ConnectAsync(SocketDevice device, CancellationToken ct = default)
    {
        Disconnect();
        _closing = false;
        LinkStateChanged?.Invoke(new LinkState.Connecting());

        _sawBytes = false;
        _sawStatus = false;
        lock (_gate) _line.Clear();

        BluetoothLEDevice? le;
        try
        {
            le = await BluetoothLEDevice.FromIdAsync(device.Id);
        }
        catch (Exception e)
        {
            Fail($"Could not open {device.Name}. {e.Message}");
            return false;
        }

        if (le is null)
        {
            Fail($"Windows could not open {device.Name}. Re-pair it in Bluetooth settings.");
            return false;
        }

        _device = le;
        le.ConnectionStatusChanged += OnConnectionStatusChanged;

        // Ask Windows to hold the radio link up. Without this the connection is
        // dropped as soon as Windows decides nothing needs it, which looks from
        // the outside like the socket randomly disappearing.
        try
        {
            _session = await GattSession.FromDeviceIdAsync(le.BluetoothDeviceId);
            _session.MaintainConnection = true;
        }
        catch (Exception)
        {
            // Not fatal - the link still works, it is just easier to lose.
        }

        // Discovery is retried because the first attempt frequently comes back
        // with only 0x1800 and 0x1801, the two services every BLE device must
        // have. That is not a device without a serial channel; it is a device
        // whose vendor service has not been enumerated yet, usually because the
        // radio connected a moment ago and GATT is still settling.
        List<BleService> described = [];
        Dictionary<Guid, List<GattCharacteristic>> handles = [];
        BleSerialProfile? profile = null;

        for (var attempt = 0; attempt < DiscoveryAttempts && profile is null; attempt++)
        {
            if (attempt > 0) await Task.Delay(1200, CancellationToken.None).ConfigureAwait(false);
            if (ct.IsCancellationRequested) break;

            described = [];
            handles = [];
            DisposeServices();

            // Uncached: a stale service list is the classic way this fails
            // silently after a module has been power-cycled or reflashed.
            var services = await le.GetGattServicesAsync(BluetoothCacheMode.Uncached);
            if (services.Status != GattCommunicationStatus.Success) continue;

            foreach (var service in services.Services)
            {
                _services.Add(service);

                var chars = await service.GetCharacteristicsAsync(BluetoothCacheMode.Uncached);
                if (chars.Status != GattCommunicationStatus.Success) continue;

                var describedCharacteristics = new List<BleCharacteristic>();
                var live = new List<GattCharacteristic>();

                foreach (var c in chars.Characteristics)
                {
                    var p = c.CharacteristicProperties;

                    describedCharacteristics.Add(new BleCharacteristic(
                        c.Uuid,
                        CanNotify: p.HasFlag(GattCharacteristicProperties.Notify)
                                   || p.HasFlag(GattCharacteristicProperties.Indicate),
                        CanWrite: p.HasFlag(GattCharacteristicProperties.Write)
                                  || p.HasFlag(GattCharacteristicProperties.WriteWithoutResponse)));

                    live.Add(c);
                }

                described.Add(new BleService(service.Uuid, describedCharacteristics));
                handles[service.Uuid] = live;
            }

            profile = BleSerialProfiles.Resolve(described);
        }

        if (profile is null)
        {
            // Naming what was actually found, rather than only what was wanted.
            // A module nobody has catalogued is exactly the case where the
            // service list is the difference between "buy another one" and
            // "add four lines to the profile table".
            var seen = string.Join(", ", described.Select(s => s.Uuid.ToString("D")));

            var onlyMandatory = described.Count > 0 && described.All(s =>
                s.Uuid.ToString("D").StartsWith("00001800-", StringComparison.OrdinalIgnoreCase) ||
                s.Uuid.ToString("D").StartsWith("00001801-", StringComparison.OrdinalIgnoreCase));

            Fail(onlyMandatory
                ? $"{device.Name} reported only the two services every BLE device has, and none of its own - after {DiscoveryAttempts} attempts. Power-cycle the socket, wait for the LED to blink fast, and try again."
                : $"{device.Name} is a Bluetooth LE device with no serial channel - nothing on it can carry the socket's data. Services found: {(seen.Length == 0 ? "none" : seen)}");

            Disconnect();
            return false;
        }

        ProfileName = profile.Name;

        if (!handles.TryGetValue(profile.Service, out var candidates))
        {
            Fail($"{device.Name} lost its serial service between discovery and use.");
            Disconnect();
            return false;
        }

        _notify = candidates.FirstOrDefault(c => c.Uuid == profile.Notify);
        _write = candidates.FirstOrDefault(c => c.Uuid == profile.Write);

        if (_notify is null || _write is null)
        {
            Fail($"{device.Name} did not offer the characteristics it advertised.");
            Disconnect();
            return false;
        }

        // Subscribing is what makes the module start pushing bytes up. Without
        // this the link is open and permanently silent.
        _notify.ValueChanged += OnValueChanged;

        var wanted = _notify.CharacteristicProperties.HasFlag(GattCharacteristicProperties.Notify)
            ? GattClientCharacteristicConfigurationDescriptorValue.Notify
            : GattClientCharacteristicConfigurationDescriptorValue.Indicate;

        GattCommunicationStatus subscribed;
        try
        {
            subscribed = await _notify.WriteClientCharacteristicConfigurationDescriptorAsync(wanted);
        }
        catch (Exception e)
        {
            Fail($"Could not subscribe to {device.Name}. {e.Message}");
            Disconnect();
            return false;
        }

        if (subscribed != GattCommunicationStatus.Success)
        {
            Fail($"{device.Name} refused a subscription ({subscribed}).");
            Disconnect();
            return false;
        }

        await SendAsync(SocketCommand.StatusNow).ConfigureAwait(false);

        var deadline = DateTime.UtcNow + HandshakeTimeout;
        while (DateTime.UtcNow < deadline)
        {
            if (_sawStatus)
            {
                LinkStateChanged?.Invoke(new LinkState.Connected(device));
                return true;
            }

            if (ct.IsCancellationRequested) break;
            await Task.Delay(100, CancellationToken.None).ConfigureAwait(false);
        }

        var reason = _sawBytes
            ? $"{device.Name} is sending data this app cannot read. If the module has a serial baud rate setting, the firmware uses 9600."
            : $"{device.Name} connected but never sent anything. Check the socket is powered, and that the module's TXD reaches Arduino pin 0.";

        Disconnect();
        Fail(reason);
        return false;
    }

    public Task SendAsync(SocketCommand command) => SendLineAsync(command.Wire());

    public async Task SendLineAsync(string line)
    {
        var write = _write;
        if (write is null) return;

        try
        {
            var buffer = CryptographicBuffer.ConvertStringToBinary(
                line + "\n",
                BinaryStringEncoding.Utf8);

            // WriteWithoutResponse where the module allows it: these modules are
            // slow to acknowledge, and a two-byte command does not need one.
            var mode = write.CharacteristicProperties
                .HasFlag(GattCharacteristicProperties.WriteWithoutResponse)
                ? GattWriteOption.WriteWithoutResponse
                : GattWriteOption.WriteWithResponse;

            await write.WriteValueAsync(buffer, mode);
        }
        catch (Exception)
        {
            // A write to a device that has just gone out of range throws. The
            // connection-status handler reports that once.
        }
    }

    public void Disconnect()
    {
        _closing = true;

        if (_notify is not null)
        {
            _notify.ValueChanged -= OnValueChanged;
            _notify = null;
        }

        _write = null;

        // Order matters on the way out too: services first, then the session,
        // then the device. Disposing the device while services are still open
        // leaves Windows holding a connection nothing can close.
        DisposeServices();

        if (_session is not null)
        {
            try { _session.MaintainConnection = false; } catch (Exception) { /* already gone */ }
            try { _session.Dispose(); } catch (Exception) { /* already gone */ }
            _session = null;
        }

        var device = _device;
        _device = null;

        if (device is not null)
        {
            device.ConnectionStatusChanged -= OnConnectionStatusChanged;
            try { device.Dispose(); } catch (Exception) { /* already gone */ }
        }

        lock (_gate) _line.Clear();
    }

    private void DisposeServices()
    {
        foreach (var service in _services)
        {
            try { service.Dispose(); } catch (Exception) { /* already gone */ }
        }

        _services.Clear();
    }

    private void OnConnectionStatusChanged(BluetoothLEDevice sender, object args)
    {
        if (_closing) return;
        if (sender.ConnectionStatus == BluetoothConnectionStatus.Disconnected)
        {
            LinkStateChanged?.Invoke(new LinkState.Failed("The link to the socket dropped."));
        }
    }

    private void OnValueChanged(GattCharacteristic sender, GattValueChangedEventArgs args)
    {
        CryptographicBuffer.CopyToByteArray(args.CharacteristicValue, out var bytes);
        if (bytes is null || bytes.Length == 0) return;

        _sawBytes = true;

        foreach (var b in bytes)
        {
            var c = (char)b;

            if (c == '\r') continue;

            if (c != '\n')
            {
                lock (_gate)
                {
                    // BLE delivers 20-byte packets, so a status line always
                    // arrives in pieces and has to be reassembled. Capping the
                    // buffer stops a module that never sends a newline from
                    // growing it without bound.
                    if (_line.Length < 128) _line.Append(c);
                }
                continue;
            }

            string complete;
            lock (_gate)
            {
                complete = _line.ToString();
                _line.Clear();
            }

            if (StatusParser.Parse(complete) is { } status)
            {
                _sawStatus = true;
                StatusChanged?.Invoke(status);
            }
        }
    }

    private void Fail(string reason) => LinkStateChanged?.Invoke(new LinkState.Failed(reason));
}

File: android/app/src/main/java/gh/group38/smartsocket/data/BleTransport.kt
package gh.group38.smartsocket.data

import android.annotation.SuppressLint
import android.bluetooth.BluetoothAdapter
import android.bluetooth.BluetoothDevice
import android.bluetooth.BluetoothGatt
import android.bluetooth.BluetoothGattCallback
import android.bluetooth.BluetoothGattCharacteristic
import android.bluetooth.BluetoothGattDescriptor
import android.bluetooth.BluetoothManager
import android.bluetooth.BluetoothProfile
import android.bluetooth.le.ScanCallback
import android.bluetooth.le.ScanResult
import android.bluetooth.le.ScanSettings
import android.content.Context
import android.os.Build
import java.util.UUID
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withTimeoutOrNull

/**
 * Bluetooth Low Energy link to a GATT serial module.
 *
 * WHY THIS EXISTS. [BluetoothTransport] speaks Classic SPP over RFCOMM, which is
 * correct and the only option for a real HC-05. The module in this project is
 * not one: Windows reported a GATT service on it, the Classic pairing never
 * produced a serial port, and the desktop app identified it as `HM-10 / FFE0` -
 * a CC2541 BLE part sold in an HC-05-shaped package under the HC-05 name.
 *
 * A BLE module has no RFCOMM socket, no SPP UUID and no serial port, so nothing
 * in the Classic transport can reach it however long you stare at the paired
 * devices list. It has to be spoken to as GATT.
 *
 * Which characteristics carry the bytes is worked out by [BleSerialProfiles], in
 * the Android-free part of the app, so this class only has to deal with the
 * radio.
 */
@SuppressLint("MissingPermission")
class BleTransport(
    private val context: Context,
    private val scope: CoroutineScope,
) : SocketTransport {

    private companion object {
        /** Standard descriptor that turns notifications on. */
        val CccdUuid: UUID = UUID.fromString("00002902-0000-1000-8000-00805f9b34fb")

        const val ConnectTimeoutMs = 15_000L

        /** Longer, because a stack-managed reconnect is meant to wait. */
        const val AutoConnectTimeoutMs = 45_000L
        const val DiscoverTimeoutMs = 10_000L
        const val HandshakeTimeoutMs = 10_000L

        /**
         * How long to wait for the stack to acknowledge one write before giving
         * up on it and letting the next through. Generous for BLE, and short
         * enough that a silent module cannot stall the queue.
         */
        const val WriteTimeoutMs = 600L

        /** A status line is longer than one BLE packet, so it arrives in pieces. */
        const val MaxLineLength = 128
    }

    private val _linkState = MutableStateFlow<LinkState>(LinkState.Idle)
    override val linkState: Flow<LinkState> = _linkState.asStateFlow()

    private val _status = MutableStateFlow(SocketStatus.UNKNOWN)
    override val status: Flow<SocketStatus> = _status.asStateFlow()

    private var gatt: BluetoothGatt? = null
    private var notifyCharacteristic: BluetoothGattCharacteristic? = null
    private var writeCharacteristic: BluetoothGattCharacteristic? = null

    private val line = StringBuilder()

    /**
     * Two separate facts, and the difference between them is the whole
     * diagnosis. Bytes with no status line means the link is fine and the two
     * ends disagree about the module's serial baud rate; no bytes at all means
     * nothing is reaching the module from the Arduino, which is a wire.
     */
    @Volatile private var sawBytes = false
    @Volatile private var sawStatus = false

    @Volatile private var connected = false
    @Volatile private var servicesFound = false
    @Volatile private var closing = false

    /** Which profile matched, once a link is up. Names the module in the UI. */
    @Volatile var profileName: String? = null
        private set

    private fun adapter(): BluetoothAdapter? =
        (context.getSystemService(Context.BLUETOOTH_SERVICE) as? BluetoothManager)?.adapter

    /**
     * BLE devices Android has actually bonded.
     *
     * USUALLY EMPTY, AND THAT IS NORMAL. Bonding is optional for BLE, and the
     * serial modules this app talks to do not ask for it - you simply connect.
     * So an HM-10 typically never appears in Android's Bluetooth settings at
     * all, never becomes bonded, and cannot be found this way. [startScan] is
     * how it is actually located; this is here for the modules that do bond.
     */
    fun pairedDevices(): List<SocketDevice> {
        val adapter = adapter() ?: return emptyList()

        return try {
            adapter.bondedDevices.orEmpty()
                .filter { it.type == BluetoothDevice.DEVICE_TYPE_LE || it.type == BluetoothDevice.DEVICE_TYPE_DUAL }
                .map {
                    SocketDevice(
                        name = it.name ?: "Unnamed",
                        address = it.address,
                        kind = LinkKind.BLE,
                    )
                }
        } catch (_: SecurityException) {
            emptyList()
        }
    }

    private var scanCallback: ScanCallback? = null

    /**
     * Looks for BLE devices on the air.
     *
     * Necessary rather than nice-to-have: a BLE serial module does not bond, so
     * it never reaches Android's paired list and a bonded-devices lookup can
     * never find it. Scanning is the only way it can be seen.
     *
     * Unfiltered on purpose. The module can be renamed by whoever configured it
     * - this one is an HM-10 calling itself HC-05 - so filtering by name would
     * hide exactly the device that needs finding. Unnamed results are kept too,
     * because a module that answers a scan without a name is still connectable
     * and its address identifies it.
     */
    fun startScan(onFound: (SocketDevice) -> Unit) {
        stopScan()

        val scanner = adapter()?.bluetoothLeScanner ?: return

        val callback = object : ScanCallback() {
            override fun onScanResult(callbackType: Int, result: ScanResult) {
                val device = result.device ?: return
                onFound(
                    SocketDevice(
                        name = device.name ?: result.scanRecord?.deviceName ?: "Unnamed",
                        address = device.address,
                        kind = LinkKind.BLE,
                        rssi = result.rssi,
                    )
                )
            }

            override fun onBatchScanResults(results: MutableList<ScanResult>) {
                results.forEach { onScanResult(0, it) }
            }
        }

        scanCallback = callback

        // Low latency: the user is watching a list and waiting. A balanced scan
        // can take tens of seconds to surface a device that is right there.
        val settings = ScanSettings.Builder()
            .setScanMode(ScanSettings.SCAN_MODE_LOW_LATENCY)
            .build()

        try {
            scanner.startScan(null, settings, callback)
        } catch (_: SecurityException) {
            scanCallback = null
        }
    }

    fun stopScan() {
        val callback = scanCallback ?: return
        scanCallback = null

        try {
            adapter()?.bluetoothLeScanner?.stopScan(callback)
        } catch (_: SecurityException) {
            // Permission revoked, or the radio went away. Either way the scan is
            // already over.
        }
    }

    /**
     * Rebuilds a dropped link, handing the waiting to the Bluetooth stack.
     *
     * autoConnect true is the difference between a link that survives a screen
     * being switched off and one that does not. With it, Android itself watches
     * for the device and reconnects whenever it comes back - at a lower duty
     * cycle, outside this process, and unaffected by Doze quietening the app.
     * It is slow to make the first connection, which is why the user-initiated
     * path does not use it.
     */
    suspend fun reconnect(device: SocketDevice): Boolean = connect(device, autoConnect = true)

    override suspend fun connect(device: SocketDevice): Boolean =
        connect(device, autoConnect = false)

    private suspend fun connect(device: SocketDevice, autoConnect: Boolean): Boolean {
        disconnect()
        closing = false
        _linkState.value = LinkState.Connecting

        sawBytes = false
        sawStatus = false
        connected = false
        servicesFound = false
        synchronized(line) { line.setLength(0) }

        val adapter = adapter() ?: run {
            fail("Bluetooth is off.")
            return false
        }

        val remote = try {
            adapter.getRemoteDevice(device.address)
        } catch (e: IllegalArgumentException) {
            fail("${device.name} is not a device this phone knows. Pair it again.")
            return false
        }

        // False for a connection the user just asked for: a direct attempt fails
        // fast, where autoConnect would sit waiting indefinitely on a socket
        // that is switched off while somebody watches a spinner.
        //
        // True when rebuilding a link that dropped on its own, where waiting is
        // exactly what is wanted and the stack does it better than this process
        // can - see reconnect().
        gatt = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            remote.connectGatt(context, autoConnect, callback, BluetoothDevice.TRANSPORT_LE)
        } else {
            remote.connectGatt(context, autoConnect, callback)
        }

        if (gatt == null) {
            fail("Could not open a link to ${device.name}.")
            return false
        }

        // A stack-managed reconnect is meant to take its time - that is the
        // point of it - so it gets a longer window than a user standing there
        // waiting would tolerate.
        val connectWindow = if (autoConnect) AutoConnectTimeoutMs else ConnectTimeoutMs

        if (!awaitFlag(connectWindow) { connected }) {
            disconnect()
            fail("${device.name} did not answer. Check the socket is powered and in range.")
            return false
        }

        if (!awaitFlag(DiscoverTimeoutMs) { servicesFound }) {
            disconnect()
            fail("${device.name} connected but would not list its services.")
            return false
        }

        val profile = BleSerialProfiles.resolve(describeServices())
        if (profile == null) {
            val seen = gatt?.services.orEmpty().joinToString { it.uuid.toString() }
            disconnect()
            fail(
                "${device.name} is a Bluetooth LE device with no serial channel - " +
                    "nothing on it can carry the socket's data. Services found: " +
                    seen.ifEmpty { "none" }
            )
            return false
        }

        profileName = profile.name

        val service = gatt?.getService(profile.service)
        notifyCharacteristic = service?.getCharacteristic(profile.notify)
        writeCharacteristic = service?.getCharacteristic(profile.write)

        val notify = notifyCharacteristic
        if (notify == null || writeCharacteristic == null) {
            disconnect()
            fail("${device.name} did not offer the characteristics it advertised.")
            return false
        }

        if (!subscribe(notify)) {
            disconnect()
            fail("${device.name} refused a subscription, so it can never send anything.")
            return false
        }

        // Ask, rather than wait. The socket publishes once a second on its own,
        // but "?" makes it answer immediately, so a good link proves itself fast.
        send(SocketCommand.STATUS_NOW)

        if (!awaitFlag(HandshakeTimeoutMs) { sawStatus }) {
            val reason = if (sawBytes) {
                "${device.name} is sending data this app cannot read. The firmware uses 9600."
            } else {
                "${device.name} connected but never sent anything. Check the socket is " +
                    "powered, and that the module's TXD reaches Arduino pin 0."
            }
            disconnect()
            fail(reason)
            return false
        }

        _linkState.value = LinkState.Connected(device)
        return true
    }

    override suspend fun send(command: SocketCommand) = sendLine(command.wire)

    /**
     * ONE WRITE AT A TIME, AND NOT NEGOTIABLE.
     *
     * Android's GATT stack allows a single outstanding operation per connection.
     * Issue a second before the first has completed and it is rejected outright
     * and silently - writeCharacteristic simply returns false, nothing is sent,
     * and nothing tells you.
     *
     * This cost real debugging: the app pushes battery, limit and charging state
     * together, and only the first of the three ever reached the socket. The LCD
     * showed the percentage and kept saying "Ready" over a charging phone,
     * looking for all the world like a firmware bug.
     */
    private val writeLock = Mutex()

    /** Completed by onCharacteristicWrite, so the next write can start. */
    @Volatile private var writeAck: CompletableDeferred<Boolean>? = null

    override suspend fun sendLine(line: String) = writeLock.withLock {
        writeLineLocked(line)
    }

    private suspend fun writeLineLocked(line: String) {
        val characteristic = writeCharacteristic ?: return
        val target = gatt ?: return

        val bytes = (line + "\n").toByteArray(Charsets.US_ASCII)
        val ack = CompletableDeferred<Boolean>()
        writeAck = ack

        // WriteWithoutResponse where the module allows it: these parts are slow
        // to acknowledge and a two-byte command does not need one.
        val type =
            if (characteristic.properties and BluetoothGattCharacteristic.PROPERTY_WRITE_NO_RESPONSE != 0) {
                BluetoothGattCharacteristic.WRITE_TYPE_NO_RESPONSE
            } else {
                BluetoothGattCharacteristic.WRITE_TYPE_DEFAULT
            }

        try {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
                target.writeCharacteristic(characteristic, bytes, type)
            } else {
                @Suppress("DEPRECATION")
                characteristic.value = bytes
                @Suppress("DEPRECATION")
                characteristic.writeType = type
                @Suppress("DEPRECATION")
                target.writeCharacteristic(characteristic)
            }
        } catch (_: SecurityException) {
            // Permission revoked mid-session. The link is about to die anyway and
            // the callback reports it once.
            writeAck = null
            return
        }

        // Wait for the stack to report the write done before releasing the lock.
        // Timed out rather than awaited forever: a module that never
        // acknowledges must not wedge every later command behind it, and the
        // next write is more useful than this one was.
        withTimeoutOrNull(WriteTimeoutMs) { ack.await() }
        writeAck = null
    }

    override fun disconnect() {
        closing = true

        val target = gatt
        gatt = null
        notifyCharacteristic = null
        writeCharacteristic = null
        connected = false
        servicesFound = false

        try {
            target?.disconnect()
            target?.close()
        } catch (_: SecurityException) {
            // Nothing useful to do; the handle is being dropped regardless.
        }

        synchronized(line) { line.setLength(0) }
        _linkState.value = LinkState.Idle
    }

    private fun describeServices(): List<BleService> =
        gatt?.services.orEmpty().map { service ->
            BleService(
                uuid = service.uuid,
                characteristics = service.characteristics.orEmpty().map { c ->
                    val p = c.properties
                    BleCharacteristic(
                        uuid = c.uuid,
                        canNotify = (p and BluetoothGattCharacteristic.PROPERTY_NOTIFY != 0) ||
                            (p and BluetoothGattCharacteristic.PROPERTY_INDICATE != 0),
                        canWrite = (p and BluetoothGattCharacteristic.PROPERTY_WRITE != 0) ||
                            (p and BluetoothGattCharacteristic.PROPERTY_WRITE_NO_RESPONSE != 0),
                    )
                },
            )
        }

    /**
     * Turning notifications on takes two steps, and missing the second is the
     * classic way a BLE link ends up connected and permanently silent: the local
     * flag only tells Android to deliver them, while the descriptor write is
     * what tells the module to send them.
     */
    private fun subscribe(characteristic: BluetoothGattCharacteristic): Boolean {
        val target = gatt ?: return false

        return try {
            if (!target.setCharacteristicNotification(characteristic, true)) return false

            val cccd = characteristic.getDescriptor(CccdUuid) ?: return false

            val value = if (characteristic.properties and
                BluetoothGattCharacteristic.PROPERTY_NOTIFY != 0
            ) {
                BluetoothGattDescriptor.ENABLE_NOTIFICATION_VALUE
            } else {
                BluetoothGattDescriptor.ENABLE_INDICATION_VALUE
            }

            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
                target.writeDescriptor(cccd, value) == BluetoothGatt.GATT_SUCCESS
            } else {
                @Suppress("DEPRECATION")
                cccd.value = value
                @Suppress("DEPRECATION")
                target.writeDescriptor(cccd)
            }
        } catch (_: SecurityException) {
            false
        }
    }

    private suspend fun awaitFlag(timeoutMs: Long, predicate: () -> Boolean): Boolean =
        withTimeoutOrNull(timeoutMs) {
            while (!predicate()) kotlinx.coroutines.delay(50)
            true
        } ?: false

    private fun fail(reason: String) {
        _linkState.value = LinkState.Failed(reason)
    }

    private fun onBytes(bytes: ByteArray) {
        if (bytes.isEmpty()) return
        sawBytes = true

        for (b in bytes) {
            val c = b.toInt().toChar()
            if (c == '\r') continue

            if (c != '\n') {
                synchronized(line) {
                    // BLE delivers 20-byte packets, so a status line always
                    // arrives in pieces and has to be reassembled. The cap stops
                    // a module that never sends a newline from growing it
                    // without bound.
                    if (line.length < MaxLineLength) line.append(c)
                }
                continue
            }

            val complete = synchronized(line) {
                val text = line.toString()
                line.setLength(0)
                text
            }

            StatusParser.parse(complete)?.let {
                sawStatus = true
                _status.value = it
            }
        }
    }

    private val callback = object : BluetoothGattCallback() {

        override fun onConnectionStateChange(g: BluetoothGatt, statusCode: Int, newState: Int) {
            when (newState) {
                BluetoothProfile.STATE_CONNECTED -> {
                    connected = true
                    try {
                        g.discoverServices()
                    } catch (_: SecurityException) {
                        // Handled by the discovery timeout.
                    }
                }

                BluetoothProfile.STATE_DISCONNECTED -> {
                    connected = false
                    if (!closing) {
                        _linkState.value = LinkState.Failed("The link to the socket dropped.")
                    }
                }
            }
        }

        override fun onServicesDiscovered(g: BluetoothGatt, statusCode: Int) {
            servicesFound = statusCode == BluetoothGatt.GATT_SUCCESS
        }

        /** Releases the next queued write. See [writeLock]. */
        override fun onCharacteristicWrite(
            g: BluetoothGatt,
            characteristic: BluetoothGattCharacteristic,
            statusCode: Int,
        ) {
            writeAck?.complete(statusCode == BluetoothGatt.GATT_SUCCESS)
        }

        @Deprecated("Kept for API < 33, which never calls the byte-array overload.")
        @Suppress("DEPRECATION")
        override fun onCharacteristicChanged(
            g: BluetoothGatt,
            characteristic: BluetoothGattCharacteristic,
        ) {
            if (Build.VERSION.SDK_INT < Build.VERSION_CODES.TIRAMISU) {
                onBytes(characteristic.value ?: ByteArray(0))
            }
        }

        override fun onCharacteristicChanged(
            g: BluetoothGatt,
            characteristic: BluetoothGattCharacteristic,
            value: ByteArray,
        ) {
            onBytes(value)
        }
    }
}
